cmd/template-deprecator/main.go

```diff
@@ -0,0 +1,88 @@
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+
+	"github.com/sirupsen/logrus"
+	prowconfig "k8s.io/test-infra/prow/config"
+	prowplugins "k8s.io/test-infra/prow/plugins"
+
+	"github.com/openshift/ci-tools/pkg/deprecatetemplates"
+)
+
+type options struct {
+	prowJobConfigDir     string
+	prowConfigPath       string
+	prowPluginConfigPath string
+	allowlistPath        string
+
+	help bool
+}
+
+func bindOptions(fs *flag.FlagSet) *options {
+	opt := &options{}
+
+	fs.StringVar(&opt.prowJobConfigDir, "prow-jobs-dir", "", "Path to a root of directory structure with Prow job config files (ci-operator/jobs in openshift/release)")
+	fs.StringVar(&opt.prowConfigPath, "prow-config-path", "", "Path to the Prow configuration file")
+	fs.StringVar(&opt.prowPluginConfigPath, "prow-plugin-config-path", "", "Path to the Prow plugin configuration file")
+	fs.StringVar(&opt.allowlistPath, "allowlist-path", "", "Path to template deprecation allowlist")
+
+	return opt
+}
+
+func (o *options) validate() error {
+	for param, value := range map[string]string{
+		"--prow-jobs-dir":           o.prowJobConfigDir,
+		"--prow-config-path":        o.prowConfigPath,
+		"--prow-plugin-config-path": o.prowPluginConfigPath,
+		"--allowlist-path":          o.allowlistPath,
+	} {
+		if value == "" {
+			return fmt.Errorf("mandatory argument %s was not set", param)
+		}
+	}
+
+	return nil
+}
+
+func main() {
+	flagSet := flag.NewFlagSet("", flag.ExitOnError)
+	opt := bindOptions(flagSet)
+	if err := flagSet.Parse(os.Args[1:]); err != nil {
+		logrus.WithError(err).Fatal("Failed to parse flags")
+	}
+
+	if opt.help {
+		flagSet.Usage()
+		os.Exit(0)
+	}
+
+	if err := opt.validate(); err != nil {
+		logrus.WithError(err).Fatal("Invalid parameters")
+	}
+
+	agent := prowplugins.ConfigAgent{}
+	if err := agent.Load(opt.prowPluginConfigPath, true); err != nil {
+		logrus.WithError(err).Fatal("Failed to read Prow plugin configuration")
+	}
+	pluginCfg := agent.Config().ConfigUpdater
+
+	prowCfg, err := prowconfig.Load(opt.prowConfigPath, opt.prowJobConfigDir)
+	if err != nil {
+		logrus.WithError(err).Fatal("failed to load Prow configuration")
+	}
+
+	enforcer, err := deprecatetemplates.NewEnforcer(opt.allowlistPath)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to initialize template deprecator")
+	}
+
+	enforcer.LoadTemplates(pluginCfg)
+	enforcer.ProcessJobs(prowCfg)
+
+	if err := enforcer.SaveAllowlist(opt.allowlistPath); err != nil {
```

Comment:
I still don't understand this. Shouldn't this be gated by a flag so we can run the tool in a presubmit to prevent ppl from adding new jobs that are not in the allowlist?

Reply:
I don't think so. Adding jobs to the allowlist is still legit thing to do for some time. The presubmit will 

1. Run the tool.
2. If fails, it means "you should do something else". There's no such check in the tool ATM, it's a followup. We would be adding more checks like this as we tighten the template world. And you are right, in that case the allowlist should not be updated on disk.
2. If is succeeds, compare the allowlist (with git). If the allowlist changed, the presubmit will also fail, but ask to include the allowlist change in the PR.

Reply:
Ah I see, I guess I expected the check to be in the tool